Range mappings: spec text for proposal

takikawa opened the pull request:
This is the actual PR for the draft range mappings proposal for detailed review. Comments and feedback welcome.

On the other PR #232, Nic raised some points about what we should do in case there are multiple mappings in one location and only one (or a subset) are range mappings.

https://tc39.es/ecma426/pr/233/

takikawa committed:
Range mappings: add spec text for decoding, parsing, and lookup (#232)

Adds a flag to Decoded Mapping Records which indicate if
a mapping is a range mapping. Populate this field based on
the decoded range mapping offsets.

nicolo-ribaudo commented on spec.emu:
```diff
@@ -947,47 +962,50 @@
             1. If _state_.[[NameIndex]] &lt; 0 or _state_.[[NameIndex]] ≥ the number of elements of _names_, optionally report an error.
             1. Else, set _name_ to _names_[_state_.[[NameIndex]]].
-          1. Let _decodedMapping_ be a new DecodedMappingRecord { [[GeneratedPosition]]: _generatedPosition_, [[OriginalPosition]]: _originalPosition_, [[Name]]: _name_ }.
+          1. <ins>Let _isRangeMapping_ be the result of performing LookupRangeMapping with arguments _rangeMappings_, _state.[[GeneratedLine]], and _state.[[MappingIndex]].</ins>
```
```suggestion
          1. <ins>Let _isRangeMapping_ be the result of performing LookupRangeMapping with arguments _rangeMappings_, _state_.[[GeneratedLine]], and _state_.[[MappingIndex]].</ins>
```

jridgewell commented on spec.emu:
```diff
@@ -682,7 +684,10 @@
         1. Let _namesField_ be GetOptionalListOfStrings(_json_, *"names"*).
-        1. Let _mappings_ be DecodeMappings(_mappingsField_, _namesField_, _sources_).
+        1. <ins>Let _rangeMappingsField_ be GetOptionalString(_json_, *"rangeMappings"*).</ins>
+        1. <ins>If _rangeMappingsField_ is *null*, set _rangeMappingsField_ to *""*.</ins>
```
Nit: just handle this in the decode function?

jridgewell commented on spec.emu:
```diff
@@ -1119,6 +1138,142 @@
+          <emu-alg>
+            1. Let _relativeIndex_ be the VLQUnsignedValue of |Vlq|.
+            1. If _relativeIndex_ is 0, then
+              1. Optionally report an error.
```
Does this mean the first mapping can never be a range mapping?

```js
"mappings": "AAAA",
"rangeMappings": "A"
```

jridgewell commented on spec.emu:
```diff
@@ -1119,6 +1138,142 @@
+            1. If _relativeIndex_ is 0, then
+              1. Optionally report an error.
+            1. Set _state_.[[MappingIndex]] to _state_.[[MappingIndex]] + _relativeIndex_.
+            1. Let _rangeMappingOffset_ be a new Decoded Range Mapping Offset Record { [[GeneratedLine]]: _state_.[[GeneratedLine]], [[MappingIndex]]: _state_.[[MappingIndex]] - 1 }.
```
Why are we dealing with 1-based indexes?

jridgewell replied:
Going though this in code, I still hate it. When I'm reading the relative offsets, I don't want to think about 1-indexing. I would rather this be 0-index, and push the complication into error checking.

This is not a strong objection though, because I can just start my relatives at `-1` and treat it like it's 0-index anyways. It just hurts my brain to reason about.

jridgewell commented on spec.emu:
```diff
@@ -931,10 +945,11 @@
           1. Set _state_.[[GeneratedColumn]] to 0.
-          1. Perform DecodeMappingsField of |LineList| with arguments _state_, _mappings_, _names_ and _sources_.
+          1. <ins>Set _state_.[[MappingIndex]] to 0.</ins>
+          1. Perform DecodeMappingsField of |LineList| with arguments _state_, _mappings_, _names_<ins>,</ins> <del>and</del> _sources_ <ins>and _rangeMappings_</ins>.
```
I think we should leave `DecodeMappingsField` alone, and instead pass `mappings` to `DecodeRangeMappings`. It allows us define error cases for 
- range mapping point to a `GenColumn` sourceless mapping
- range mapping points to undefined mapping (line too large or index too large)

takikawa committed:
Range mappings: quick fixes for review feedback (#234)

* Address some quick fixes from review

* Move null check into decode operation

* Return an empty List on null range mappings

takikawa committed:
Range mappings: refactor decoding out of mapping decoding (#235)

* Refactor range mapping decoding based on feedback

  * Range mapping decoding happens separate of and after mapping decoding now
  * Mappings are updated in-place to add range mapping info

* Continue if mapping is NOT-FOUND in range mapping update

jridgewell commented on spec.emu:
```diff
@@ -1119,6 +1128,168 @@
+          </h1>
+          <dl class="header"></dl>
+          <emu-alg>
+            1. Let _currentLine_ be *null*.
```
Nit: `-1`, so we have a consistent number type?

takikawa committed:
Range mappings: Change index encoding to start at 0 (#252)

* Change range mapping index encoding to start at 0

Currently "BB" means the first and second mappings are
range mappings. Instead, we'll have "AB" encode the same.
"BA" will be an (optional) error because a duplicate
index is encoded.

* Fix detection of duplicate range mapping offsets

Disallows a range mapping string like "AA" that
wasn't covered by the previous attempt.

* Add missing empty RangeMappingLine case

* Minor fix: replace not equal sign with "is not"
